openeo-geotrellis/src/test/scala/org/openeo/geotrellis/layers/FileLayerProviderTest.scala

```diff
@@ -981,4 +1000,98 @@ class FileLayerProviderTest {
+      Extent(703109 - 100, 5600100, 709000, 5610000 - 100),
+      LocalDate.of(2023, 4, 5),
+    )
+    val cubeSpatial = layer.toSpatial()
+    cubeSpatial.writeGeoTiff("tmp/testPixelValueOffsetNeededCorner.tiff")
+    val arr = cubeSpatial.collect().array
+    assertTrue(isNoData(arr(1)._2.toArrayTile().band(0).get(162, 250)))
+    assertEquals(224, arr(0)._2.toArrayTile().band(0).get(0, 0), 1)
+  }
+
+  @Test
+  def testPixelValueOffsetNeededDark(): Unit = {
+    // This will cover an area where pixels go under 0
+    val layer = testPixelValueOffsetNeeded(
+      "/org/openeo/geotrellis/testPixelValueOffsetNeededDark.json",
+      Extent(755380, 5622042, 756755, 5623482),
+      LocalDate.of(2023, 1, 17),
+    )
+    val cubeSpatial = layer.toSpatial()
+    cubeSpatial.writeGeoTiff("tmp/testPixelValueOffsetNeededDark.tiff")
+    val band = cubeSpatial.collect().array(0)._2.toArrayTile().band(0)
+
+    // Snippet to find relevant pixels to extend test case:
+    //    1 until band.cols foreach (x => 1 until band.rows foreach (y => if (band.get(x, y) < -999)
+    //      println("point (" + x + "," + y + ") " + band.get(x, y))
+    //    ))
+    assertEquals(888, band.get(0, 0), 1)
+    assertEquals(-582, band.get(133, 151), 1)
+  }
+
+  def testPixelValueOffsetNeeded(
+                                  jsonPath: String,
+                                  extent: Extent,
+                                  localDate: LocalDate,
+                                ): MultibandTileLayerRDD[SpaceTimeKey] = {
+    val bandNames = Collections.singletonList("B04")
+    val srs32631 = "EPSG:32631"
+    val projected_polygons_native_crs = ProjectedPolygons.fromExtent(extent, srs32631)
+
+    val jsonPathFull = getClass.getResource(jsonPath)
+    val fileSource = Source.fromURL(jsonPathFull)
+    var txt = try fileSource.mkString
+    finally fileSource.close()
+    val basePath = new java.io.File(jsonPathFull.getFile).getParent
+    // Use artifactory to avoid heavy git repo
+    val basePathArtifactory = "https://artifactory.vgt.vito.be/testdata-public"
+
+    for (rest <- Seq(
+      "/eodata/Sentinel-2/MSI/L2A/2023/01/17/S2B_MSIL2A_20230117T104259_N0509_R008_T31UGS_20230117T120337.SAFE/manifest.safe",
+      "/eodata/Sentinel-2/MSI/L2A/2023/01/17/S2B_MSIL2A_20230117T104259_N0509_R008_T31UGS_20230117T120337.SAFE/GRANULE/L2A_T31UGS_A030636_20230117T104258/IMG_DATA/R10m/T31UGS_20230117T104259_B04_10m.jp2",
+      "/eodata/Sentinel-2/MSI/L2A/2023/04/05/S2A_MSIL2A_20230405T105031_N0509_R051_T31UFS_20230405T162253.SAFE/manifest.safe",
+      "/eodata/Sentinel-2/MSI/L2A/2023/04/05/S2A_MSIL2A_20230405T105031_N0509_R051_T31UFS_20230405T162253.SAFE/GRANULE/L2A_T31UFS_A040660_20230405T105026/IMG_DATA/R10m/T31UFS_20230405T105031_B04_10m.jp2",
+    )) {
+      val jp2File = new File(basePath, rest)
+      if (!jp2File.exists()) {
+        println("Copy from artifactory to: " + jp2File)
+        FileUtils.copyURLToFile(new URL(basePathArtifactory + rest), jp2File)
```

Comment:
This function is runned 2x, totalling 4min in the test suite.  
@bossie , do you think there is an easy way to make this download faster? By cropping or lossy compressing the jp2 for example

Reply:
Those are some good options but they still require some effort.

BTW I used test files in the past but stored them on NFS instead of in Artifactory; I don't think there's much to be gained there though.

https://github.com/Open-EO/openeo-geotrellis-extensions/blob/68143219e6882dffd0d05a46aaf29e530c0b93a5/geotrellis-sentinelhub/src/test/scala/org/openeo/geotrellissentinelhub/PyramidFactoryTest.scala#L325C4-L326

Ideally you generate the jp2 yourself in code with exactly the pixels that you need for the test and avoid the download altogether.

Reply:
I'll leave as it is then. I tried to refer to them by URL, hoping that they would only be partially downloaded, but I got gdal related errors and aborted that plan